### ISSUE TITLE
Upgrade to 2.X results in many "TS2589: Type instantiation is excessively deep and possibly infinite." errors

### ISSUE THREAD

**Issue by svrnwnsch** [Vincit/objection.js#1668]
We upgraded recently our typescript project from 1.6.11 to 2.1.2. Since then we have many Typescript errors: "TS2589: Type instantiation is excessively deep and possibly infinite."

These errors mainly occur in place where we call. `await object.$query().update({...})` or `await Class.query().select('*', knex.raw('....'))`

I guess this has somewhat to do with the new types introduced in version 2. In general the functionality seems to work as before. So we are just ignoring this specific error.
Maybe this issue is somehow related to: https://github.com/microsoft/TypeScript/issues/34933?

We use Typescript: 3.7.5

**Comment by petrdsvoboda** [comment]
I encountered the same problem. Also in typescript 3.7.
I can't provide any small reproduction case as it appears to be a problem only in larger codebases.

**Comment by koskimas** [comment]
As this is a typescript bug, there's not much I can do. I could try to find a workaround if I had a way to reproduce this.

**Comment by svrnwnsch** [comment]
Looking at [this comment](https://github.com/microsoft/TypeScript/issues/34933#issuecomment-552500444) I guess that the problem could come from something like, defining two models witch have a reference to each other and if the models look something like:

```
export class ModelA extends Model {
   modelBs: ModelB[]
}
export class ModelB extends Model {
   modelAs: ModelA[]
}
```

If I have some more time I might try to test if such a case will create this issue.

**Comment by petrdsvoboda** [comment]
I also think this might be the case.

I tried using the new `import type` syntax from typescript 3.8.
```ts
import type ModelA from './ModelA'
export class ModelB extends Model {
   modelAs: ModelA[]
}
```
Trying if it would help if the references are set only to a type. But that didn't work for me.

**Comment by willsoto** [comment]
We are having a similar issue in our codebase that only appeared after the 2.x upgrade. The new typings are much improved, but I wonder if it's something related to [these typings](https://github.com/Vincit/objection.js/blob/9697b276de4461c6fb954c033ed4708742ec46eb/typings/objection/index.d.ts#L230-L273) that is causing these issues. 

It could very well be a TypeScript issue, as there is an open bug there, but I wonder if there is something that can be done here to help mitigate the issue? 

I did come across [this line from another ORM](https://mikro-orm.io/docs/entity-manager/#mitigating-type-instantiation-is-excessively-deep-and-possibly-infinitets2589-error). Wonder if it could help here?

**Comment by petrdsvoboda** [comment]
There doesn't seem to be any response on typescript side.

Would it possible to change the objection.js typings? This issue is blocking me to migrate to the objection v2 without major rewrite. I think that willsoto is right.

> We are having a similar issue in our codebase that only appeared after the 2.x upgrade. The new typings are much improved, but I wonder if it's something related to [these typings](https://github.com/Vincit/objection.js/blob/9697b276de4461c6fb954c033ed4708742ec46eb/typings/objection/index.d.ts#L230-L273) that is causing these issues.

Unfortunately I am not sure if I will be able to fix it myself, but I found a PR fixing this issue in other codebase https://github.com/davidkpiano/xstate/pull/1103.

**Comment by willsoto** [comment]
@petrdsvoboda I wouldn't so sure. [This issue](https://github.com/microsoft/TypeScript/issues/33460) was just closed less than a week ago on the TypeScript. Read the last two comments.

**Comment by petrdsvoboda** [comment]
I can confirm that for me the issue is resolved on typescript `3.9.1-rc`.

**Comment by koskimas** [comment]
I believe this is now fixed. Please open a new issue if not.
